src/Bounds.cpp

```diff
@@ -1055,11 +1055,31 @@ class Bounds : public IRVisitor {
                             equiv.accept(this);
                         }
                     } else if (op->is_intrinsic(Call::shift_right)) {
-                        if (a_interval.has_lower_bound() && b_interval.has_upper_bound()) {
-                            interval.min = a_interval.min >> b_interval.max;
-                        }
-                        if (a_interval.has_upper_bound() && b_interval.has_lower_bound()) {
-                            interval.max = a_interval.max >> b_interval.min;
+                        // shift_right(a, b) is UB for b < 0, so only try to improve on bounds-of-type
+                        // if we can prove b >= 0.
+                        if (b_interval.is_bounded() && (t.is_uint() || can_prove(b_interval.min >= 0))) {
+                            if (a_interval.has_lower_bound()) {
+                                if (t.is_uint() || can_prove(a_interval.min >= 0)) {
+                                    interval.min = a_interval.min >> b_interval.max;
+                                } else {
+                                    // if a < 0, the smallest value will be a >> b.min
+                                    // if a > 0, the smallest value will be a >> b.max
+                                    interval.min = a_interval.min >> select(a_interval.min < 0,
```

Comment:
This will probably be accepted more cleanly by other analysis passes as:

``` min(a_interval.min >> b_interval.min, a_interval.min >> b_interval.max)```

Reply:
min() is always better than select()